### PR TITLE
Add end-to-end authentication and authorization

### PR DESCRIPTION
**Pull request by marcellp**
# Background
#### Link to issue
https://biomage.atlassian.net/browse/BIOMAGE-860

#### Link to staging deployment URL
https://ui-marcellp-auth-e2e.scp-staging.biomage.net/experiments/marcellp-auth-e2e-e52b39624588791a7889e39c617f669e/data-processing

#### Links to any Pull Requests related to this
https://github.com/biomage-ltd/ui/pull/250

#### Anything else the reviewers should know about the changes here
N/A

# Changes
### Code changes
Added a proper authentication and authorization middleware for Socket.IO and Express in the API.

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [x] Unit tests written
- [x] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [x] Relevant Github READMEs updated
- [x] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/utils/errors.js
Comment (marcellp) [inline src/utils/errors.js]: This is very annoying given 401 is called "Unauthorized", but it is actually for unauthenticated requests.

File: tests/api/routes/experiment.test.js
Comment (marcellp) [inline tests/api/routes/experiment.test.js:6]: Mock out the middleware for the standard unit test. We will test authorization working end-to-end in a separate unit test.

File: src/utils/authMiddlewares.js
Comment (marcellp) [inline src/utils/authMiddlewares.js:38]: JWT tokens are susceptible for downgrade attacks if the algorithm used to sign the token is not specified. AWS only returns `RS256` signed tokens so we explicitly specify this requirement.

File: src/utils/authMiddlewares.js
Comment (marcellp) [inline src/utils/authMiddlewares.js:47]: JWT tokens are basically JSON files that are signed using a key. Anyone can create one using an encryption key, so we need to make sure that the issuer in the token is correct. We verify that the signed JWT includes our own user pool.

File: src/utils/authMiddlewares.js
Comment (marcellp) [inline src/utils/authMiddlewares.js:52]: We extract the key ID here that was used to sign the JWT.

File: src/utils/authMiddlewares.js
Comment (marcellp) [inline src/utils/authMiddlewares.js:63]: We do some caching to ensure that we don't make HTTP calls every time a user wants to do something with the API.

File: src/utils/authMiddlewares.js
Comment (marcellp) [inline src/utils/authMiddlewares.js:79]: The public key to verify the signature on the JWT comes from a pre-determined URL linked to the issuer. 

Verifying the issuer is therefore even more important as not doing so can cause anyone to create an arbitrary key, and set the issuer to a web server they control.

File: src/utils/authMiddlewares.js
Comment (marcellp) [inline src/utils/authMiddlewares.js:43]: This is okay as we don't always need users to be authenticated to perform an action (e.g. a health check does not require authentication). The authenticated claim, if successful, gets saved in the request under `req.user`, so during authorization we can check if the `user` parameter is actually present. If not, the user was not authenticated.

File: src/utils/authMiddlewares.js
Comment (marcellp) [inline src/utils/authMiddlewares.js:90]: Similar functionality to the express middleware, but using a different package.

File: src/utils/authMiddlewares.js
Comment (marcellp) [inline src/utils/authMiddlewares.js]: As said above, if the user is not authenticated, `req.user` will be none, so we just check it in the authorization and return a HTTP 401 in the authorization middleware if that is the case.

File: src/utils/authMiddlewares.js
Comment (cosa65) [inline src/utils/authMiddlewares.js]: Rename to unauthenticated?

File: src/utils/authMiddlewares.js
Comment (cosa65) [inline src/utils/authMiddlewares.js]: based on authorization status*

File: src/utils/authMiddlewares.js
Comment (cosa65) [inline src/utils/authMiddlewares.js:110]: Does it make sense to perhaps factor out into a different function the fetching of the key? Since it is being used both here and in the other middleware.

File: api.code-workspace
Comment (ivababukova) [inline api.code-workspace]: why did you remove this?

Comment (marcellp) [reply api.code-workspace]: it is added back into a separate settings.json file so you can open one workspace with all the projects.

File: src/utils/authMiddlewares.js
Comment (marcellp) [reply src/utils/authMiddlewares.js:110]: It would be good however the functions themselves are quite different. I could factor out the actual fetch call, but even then because we have two different types of caching being used I’ve found it hard to abstract it away properly

File: api.code-workspace
Comment (ivababukova) [reply api.code-workspace]: ok, I see. Upon merging this, we will need to tell the rest of the team, because otherwise this may cause confusion -- just don't forget to post on slack #general telling people that this has been changed and how to use the settings.json file

File: src/utils/authMiddlewares.js
Comment (cosa65) [reply src/utils/authMiddlewares.js:110]: I was thinking mainly handling the lines 90, 91 and 92 in the new function and returning a promise to separate caching from fetching, but if it has the potential to cause issues I think its ok like this.

File: src/utils/authMiddlewares.js
Comment (marcellp) [reply src/utils/authMiddlewares.js]: Fixed, it was a typo.

File: src/utils/authMiddlewares.js
Comment (marcellp) [reply src/utils/authMiddlewares.js]: Fixed.

**Comment by marcellp** [comment]
The authorization middleware will be disabled before merging if this PR is approved. This is because users do not have accounts yet. There is a need from @vickymorrison to have a release for a user onboarding soon so we can't merge this in as is and give users accounts until the next release.

**Comment by marcellp** [comment]
> The authorization middleware will be disabled before merging if this PR is approved. This is because users do not have accounts yet. There is a need from @vickymorrison to have a release for a user onboarding soon so we can't merge this in as is and give users accounts until the next release.

This is not true anymore, see https://this-is-biomage.slack.com/archives/C014YMUT6GN/p1621021422046600. This can be merged once reviewers approve.

**Comment by cosa65** [comment]
I tried to use the staging link but no luck for now, when I first try to enter the staging link the UI shows the message for an error 500 (this part does not happen when using incognito in chrome).

If I go to data management I can see I am logged in with your account.
If instead of clicking data management I refresh the page I get an error 503, and I seem to keep getting that error whenever I try again, even logging in again in incognito.

Looking at the api in lens, there are a bunch of "The user is not authenticated" messages (which I believe mostly were triggered by my attempts).

File: src/api/events/index.js
Comment (ivababukova) [inline src/api/events/index.js]: delete the second `the user`

**Review by ivababukova — COMMENTED** [review]
Code-wise looks good, but before we merge it, we have to make sure that all staging and production DynamoDB experiments table entries contain Marcell's cognito token. Or we will loose access to some experiments after merging this.

**Review by ivababukova — APPROVED** [review]

**Pull request merged**
